test/unit-test/test_derivation.ml

```diff
@@ -188,44 +183,43 @@ let%expect_test "derivation" =
         (Subtype (TySingleton (Number 57)) (TyVar b))
         Empty)))) |}];
 
-  print Context.empty (App (Abs (["x"], Var "x"), [Constant (Number 42)]));
+  print Context.empty (App (Abs (["X"], Var "X"), [Constant (Number 42)]));
   [%expect {|
     (Ok (
-      (TyVar e)
+      (TyVar d)
       (Conj (
-        (Eq (TyVar b) (TyFun ((TyVar c)) (TyVar d)))
+        (Eq
+          (TyFun ((TyVar a)) (TyVar a))
+          (TyFun ((TyVar b)) (TyVar c)))
         (Subtype
-          (TyVar e)
-          (TyVar d))
-        (Eq (TyVar b) (TyConstraint (TyFun ((TyVar a)) (TyVar a)) Empty))
-        (Subtype (TySingleton (Number 42)) (TyVar c))
-        Empty)))) |}];
+          (TyVar d)
+          (TyVar c))
+        Empty
+        (Subtype (TySingleton (Number 42)) (TyVar b))
+        Empty))))
+    |}];
 
-  print Context.empty (App (Abs (["x"; "y"], Var "x"), [Constant (Number 42); Constant (Number 57)]));
+  print Context.empty (App (Abs (["X"; "Y"], Var "X"), [Constant (Number 42); Constant (Number 57)]));
   [%expect {|
     (Ok (
-      (TyVar g)
+      (TyVar f)
       (Conj (
         (Eq
-          (TyVar c)
           (TyFun
-            ((TyVar d)
-             (TyVar e))
-            (TyVar f)))
+            ((TyVar a)
+             (TyVar b))
+            (TyVar a))
+          (TyFun
+            ((TyVar c)
+             (TyVar d))
+            (TyVar e)))
         (Subtype
-          (TyVar g)
-          (TyVar f))
-        (Eq
-          (TyVar c)
-          (TyConstraint
-            (TyFun
-              ((TyVar a)
-               (TyVar b))
-              (TyVar a))
-            Empty))
-        (Subtype (TySingleton (Number 42)) (TyVar d))
+          (TyVar f)
+          (TyVar e))
         Empty
-        (Subtype (TySingleton (Number 57)) (TyVar e))
+        (Subtype (TySingleton (Number 42)) (TyVar c))
+        Empty
+        (Subtype (TySingleton (Number 57)) (TyVar d))
```

Comment:
I checked this derivation with paper.
![derivation tree](https://pbs.twimg.com/media/DuDC9N2VsAAOlL3.jpg)